src/libra_py/pdos.py

```diff
@@ -384,4 +384,154 @@ def libra_pdos(_emin, _emax, _de, projections, prefix, outfile, Nel, do_convolve
+        
+        coef (float): The coefficient multiplied in Gaussian function.
+        
+        npoints (int): The number of points used in convolution.
+        
+        energy_conversion (float): The energy conversion unit from Hartree. For example 27.211386 is
+                                   for unit conversion from Hartree to eV.
+				   
+        angular_momentum_cols (list): The angular momentum columns in the *.pdos files produced by CP2K.
+	
+    Returns:
+	
+        energy_grid (numpy array): The energy grid points vector.
+		
+        convolved_pdos (numpy array): The convolved pDOS vector.
+		
+        homo_energy (float): The average HOMO energy.
+		
+    """
+
+    # Opening the file
+    file = open(cp2k_log_file,'r')
+    lines = file.readlines()
+    file.close()
+    
+    # Lines with 'DOS'
+    lines_with_dos = []
+    
+    # Finding the lines with 'DOS'
+    for i in range(0,len(lines)):
+        if 'DOS'.lower() in lines[i].lower().split():
+            lines_with_dos.append(i)
+    
+    # Finding the first and last index of PDOS for each time step
+    if len(lines_with_dos)==1:
+        # First index
+        first_index = 2
+        # Last index
+        last_index = int(lines[len(lines)-1].split()[0])
+    elif len(lines_with_dos)>1:
+        # First index
+        first_index = 2
+        # Last index
+        last_index = int(lines_with_dos[1]-1)
+    
+    # Find the number of columns in the PDOS file showing the number 
+    # of orbital components, energy, and occupation column.
+    num_cols = len(lines[first_index].split())
+    
+    # Number of energy levels considered for PDOS
+    num_levels = last_index - first_index + 1
+
+    
+    # Finding the homo and lumo energy level by appending the 
+    # pdos numerical values of unoccupied states only
+    pdos_unocc = []
+    # Energy levels
+    energy_levels = []
+    for i in range(first_index, last_index + 1):
+        energy_levels.append(float(lines[i].split()[1])*energy_conversion)
+        if float(lines[i].split()[2])==0:
+            pdos_unocc.append(i)
+    # LUMO energy level
+    lumo_level = int(lines[min(pdos_unocc)].split()[0])
+    # HOMO energy level
+    homo_level = lumo_level-1
+    # HOMO energy
+    homo_energy = float(lines[homo_level].split()[1])*energy_conversion
+    # Minimum energy level
+    min_energy = float(lines[first_index].split()[1])*energy_conversion
+    # Maximum energy level
+    max_energy = float(lines[last_index].split()[1])*energy_conversion
+    
+    
+    # Now we make an equispaced energy vector from min_energy ad max_energy with npoints.
+    energy_grid = np.linspace( min_energy-2, max_energy+2, npoints )
+    energy_grid = np.array(energy_grid)
+    
+    
+    # Appending the energy lines with their component densities of states
+    energy_lines = []
+    for i in range( time_step * ( num_levels + 2 ) + 2, ( time_step + 1 ) * ( num_levels + 2 ) ):

```

Comment:
make the inputs of the range function the pre-computed variables - don't do the calculations in the range function itself